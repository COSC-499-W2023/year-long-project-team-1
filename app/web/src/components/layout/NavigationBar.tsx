/*
 * Copyright [2023] [Privacypal Authors]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use client";

import {
  Masthead,
  MastheadMain,
  MastheadBrand,
  MastheadContent,
} from "@patternfly/react-core";
import Link from "next/link";
import PrivacyPalLogo from "./PrivacyPalLogo";
import { LoginLogout } from "@components/auth/link/LoginLogout";
import { User } from "next-auth";
import ProfilePicture from "./ProfilePicture";

const styles: { [key: string]: React.CSSProperties } = {
  navbar: {
    width: "100%",
  },
  main: {
    display: "flex",
    alignItems: "center",
    gap: "1rem",
    color: "var(--pf-global--Color--200)",
  },
  logo: {
    margin: "0.5rem 0",
  },
  content: {
    display: "flex",
    gap: "1rem",
    justifySelf: "flex-end",
  },
  link: {
    color: "var(--pf-global--Color--200)",
  },
};

interface NavigationBarProps {
  user?: User;
  className?: string;
  style?: React.CSSProperties;
}

export default function NavigationBar({
  user,
  className,
  style,
}: NavigationBarProps) {
  // TODO: enable masthead toggle
  return (
    <Masthead style={{ ...styles.navbar, ...style }} className={className}>
      {/* <MastheadToggle>
        <Button
          variant="plain"
          onClick={() => {}}
          aria-label="Global navigation"
        >
          <BarsIcon />
        </Button>
      </MastheadToggle> */}
      <MastheadMain style={styles.main}>
        <MastheadBrand component={(props) => <Link {...props} href="/" />}>
          <PrivacyPalLogo style={styles.logo} w={48} h={48} dark={false} />
        </MastheadBrand>
        <Link href="/" style={styles.link}>
          Home
        </Link>
      </MastheadMain>
      <MastheadContent style={styles.content}>
        <LoginLogout user={user} />
        {user ? (
          <ProfilePicture
            tooltip={`Logged in as ${user.username}`}
            user={user}
          />
        ) : null}
      </MastheadContent>
    </Masthead>
  );
}
