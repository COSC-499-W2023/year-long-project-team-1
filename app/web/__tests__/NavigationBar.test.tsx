/*
 * Copyright [2023] [Privacypal Authors]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { User } from "next-auth";
import { Role } from "@prisma/client";
import NavigationBar from "@components/layout/NavigationBar";

describe("NavigationBar", () => {
  const testUser: User = {
    id: 1,
    firstName: "Test",
    lastName: "User",
    username: "testuser",
    role: Role.CLIENT,
    email: "testuser@example.com",
    phone_number: "+1234567890",
  };

  it("matches snapshot", () => {
    const wrapper = render(<NavigationBar user={testUser} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("generates an HTML header element", () => {
    render(<NavigationBar user={testUser} />);
    expect(screen.getByRole("banner")).toBeInTheDocument();
  });

  it("renders the site logo", () => {
    render(<NavigationBar user={testUser} />);
    const logo = screen.getByAltText("PrivacyPal Shield Logo");
    expect(logo).toBeInTheDocument();
  });

  it("renders a link to /", () => {
    render(<NavigationBar user={testUser} />);
    const link = screen.getByRole("link", { name: "Home" });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute("href", "/");
  });

  it("shows a login link when no user is provided", () => {
    render(<NavigationBar />);
    const link = screen.getByRole("button", { name: "Sign in" });
    expect(link).toBeInTheDocument();
  });

  it("shows a logout link when a user is provided", () => {
    render(<NavigationBar user={testUser} />);
    const link = screen.getByRole("button", { name: "Sign out" });
    expect(link).toBeInTheDocument();
  });
});
