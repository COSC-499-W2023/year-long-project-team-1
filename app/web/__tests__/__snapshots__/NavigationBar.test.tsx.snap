// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="pf-v5-c-masthead pf-m-display-inline-on-md"
        style="width: 100%;"
      >
        <div
          class="pf-v5-c-masthead__main"
        >
          <a
            class="pf-v5-c-masthead__brand"
            href="/"
            style="display: flex; align-items: center; gap: 1rem;"
          >
            <img
              alt="PrivacyPal Shield Logo"
              data-nimg="1"
              decoding="async"
              height="48"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
              style="color: transparent; margin: 0.5rem 0px;"
              width="48"
            />
          </a>
          <a
            href="/"
          >
            Home
          </a>
        </div>
        <div
          class="pf-v5-c-masthead__content"
          style="display: flex; gap: 1rem; justify-self: flex-end;"
        >
          <button
            aria-disabled="false"
            class="pf-v5-c-button pf-m-primary"
            data-ouia-component-id="OUIA-Generated-Button-primary-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            Sign out
          </button>
          <a
            href="/user/dashboard"
            style="display: flex; flex-direction: column; align-items: center; margin: 0.5rem 0px; text-decoration: none;"
            title="Logged in as testuser"
          >
            <img
              alt="Profile picture"
              class="pf-v5-c-avatar pf-m-md"
              src="/img.jpg"
              style="cursor: pointer;"
            />
            <span
              style="font-size: 0.75rem; font-weight: bold;"
            >
              testuser
            </span>
            <span
              style="font-size: 0.66rem; font-weight: lighter; text-transform: capitalize;"
            >
              CLIENT
            </span>
          </a>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="pf-v5-c-masthead pf-m-display-inline-on-md"
      style="width: 100%;"
    >
      <div
        class="pf-v5-c-masthead__main"
      >
        <a
          class="pf-v5-c-masthead__brand"
          href="/"
          style="display: flex; align-items: center; gap: 1rem;"
        >
          <img
            alt="PrivacyPal Shield Logo"
            data-nimg="1"
            decoding="async"
            height="48"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
            style="color: transparent; margin: 0.5rem 0px;"
            width="48"
          />
        </a>
        <a
          href="/"
        >
          Home
        </a>
      </div>
      <div
        class="pf-v5-c-masthead__content"
        style="display: flex; gap: 1rem; justify-self: flex-end;"
      >
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-primary"
          data-ouia-component-id="OUIA-Generated-Button-primary-1"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          type="button"
        >
          Sign out
        </button>
        <a
          href="/user/dashboard"
          style="display: flex; flex-direction: column; align-items: center; margin: 0.5rem 0px; text-decoration: none;"
          title="Logged in as testuser"
        >
          <img
            alt="Profile picture"
            class="pf-v5-c-avatar pf-m-md"
            src="/img.jpg"
            style="cursor: pointer;"
          />
          <span
            style="font-size: 0.75rem; font-weight: bold;"
          >
            testuser
          </span>
          <span
            style="font-size: 0.66rem; font-weight: lighter; text-transform: capitalize;"
          >
            CLIENT
          </span>
        </a>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
